Added languages table and relationship to repos

mnorth719 opened the pull request:
@thingdeux Updated based on feedback. Didn't know about that force join, thats pretty sweet! Thanks for the feedback, I promise I'll do my best to stop forcing type safety :)

mnorth719 committed:
Added languages table and relationship to repos

mnorth719 committed:
updated languages per repo to return list

thingdeux commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 .idea/
 *.db
+*.pyc
```
Attaboy!

thingdeux commented on src/database/models/language.py:
```diff
@@ -0,0 +1,24 @@
+
+
+# ORM Class for Repositories
+class Language(base_model()):
```
I don't disagree with your making these many-to-many relationships for language re-use .. it's a bit more expensive to do but it's right.  Can you make sure however to put a unique constraint on name here so there's only ever 1 type of a given language.

thingdeux commented on src/database/models/repository.py:
```diff
@@ -26,4 +28,5 @@ class Repository(base_model()):
-    is_private = Column(Boolean)    # Flag for whether or not the repo is private
\ No newline at end of file
+    is_private = Column(Boolean)    # Flag for whether or not the repo is private
+    languages = relationship("Language", secondary=repo_language_table, back_populates='repositories')
```
Might be useful for force 'joined' eager style here.
http://docs.sqlalchemy.org/en/latest/orm/loading_relationships.html#joined-eager-loading
http://docs.sqlalchemy.org/en/latest/orm/loading_relationships.html#configuring-loader-strategies-at-mapping-time

Because you know for a fact that you'll always need the languages along with the repo (unless that changes later) it makes since to always join this data in one select statement than do one query to get the repo and another to join.

thingdeux commented on src/database/models/repository.py:
```diff
@@ -4,10 +4,12 @@
 
 # ORM Class for Repositories
 class Repository(base_model()):
+    from database.models.language import repo_language_table
```
You should be able to import this outside of the class scope - you only reference repository in a string instead of directly - so you shouldn't run into any circular import issues.

thingdeux commented on src/models/code_language.py:
```diff
@@ -0,0 +1,20 @@
+
+class CodeLanguage(LanguageStorable):
+    def __init__(self, dictionary):
+        if isinstance(dictionary, dict):
```
I don't want you to get in the habit of checking to make sure everything passed into a constructor is what it should be.  That's a bit non-pythonic - I'd like to see you maybe follow the pattern of type hinting the parameter and then just assuming the person who uses the constructor will use it responsibly.

thingdeux commented on src/database/language_service.py:
```diff
@@ -0,0 +1,41 @@
+
+    @staticmethod
+    def get_languages_for_repo(repo_id: int) -> [Language]:
+        if isinstance(repo_id, int):
```
I'd say remove this isinstance - your type hint is enough.  If someone misuses this it'll crash and then they'll change it.

thingdeux commented on src/database/repo_service.py:
```diff
@@ -17,6 +19,13 @@ def insert_new_code_repo(code_repo: RepoStorable):
                     session = db_manager.get_session()
+
+                    # TODO dont hardcode this to all languages
+                    languages = session.query(Language).all()
```
As a suggestion for how to do this when the time comes....instead of hitting the DB and querying each language on the fly - I'd say hit the DB once - map the results to a dictionary keyed off of the language with a value of the db object and then as you go through the associated language apply as needed:

Ex:
```
languages = ls_action.get_all_languages()
languages_mapping_dictionary =  {key: language.name.lower() for language in languages}
```

Then as you create or update your repos you can do this
```
for repo in repos:
     languages_to_relate = some_mapper_for_languages[repo.id]
     for related_language in languages_to_relate:
         repo.languages.append(languages_mapping_dictionary[language_name])
```
1 SQL query gets you all that you need.

thingdeux reviewed (changes requested):
Looks good!

I'd say the only thing that's a must before merge is unique constraint on language name.  Also I'd make sure on insert to language that any string is lowercased.

mnorth719 committed:
Updated based on PR feedback

- removed non-pythonic type checks
- changed repo/language relationship to force 'joined' eager style
- enforce name uniqueness of languages

thingdeux reviewed (approved):
FREEEEDOM!
#Pythonic
